Add class to statically get asset URLs similar to Router::url() for routes.

ADmad opened the pull request:
Refs https://www.reddit.com/r/cakephp/comments/alnoz8/how_to_generate_an_asseturl_with_the_urlhelper_in/

ADmad committed:
Add Asset class to statically generate asset URLs.

This allows generating asset URLs for cases like generating JSON response
without template rendering.

ADmad committed:
Add helper method to get request's webroot.

ADmad commented:
Once this is approved I can retrofit `UrlHelper` to use this class in it's methods and avoid code duplication.

One thing to note is this class only does URL encoding and doesn't do any HTML entities encoding. I believe the latter should be done only when this class is used through helpers where are meant to generate HTML.

markstory commented on src/View/Asset.php:
```diff
@@ -0,0 +1,336 @@
+/**
+ * Class for generating asset URLs.
+ */
+class Asset
```
Should this be in Routing package?

ADmad replied:
Probably, I wasn't too sure of the location.

markstory replied:
Routing seems like a better location for this if we want to make a general asset URL building class.

markstory commented on src/View/Asset.php:
```diff
@@ -0,0 +1,336 @@
+     */
+    protected static function inflectThemeName(string $name): string
+    {
+        return Inflector::underscore($name);
```
Do we need this additional function call/indirection?

ADmad replied:
I copied this from UrlHelper, I can remove it.

markstory reviewed (commented):
If we add this class would we replace the guts of UrlHelper with this?

ADmad commented:
> If we add this class would we replace the guts of UrlHelper with this?

Yes

ADmad committed:
Move Asset class from View to Routing.

ADmad committed:
Use common helper method to inflect plugin / theme names.

ADmad committed:
Update UrlHelper to use Asset class.

ADmad committed:
Rename Asset methods.

ADmad committed:
Allow setting the inflection type to use in Asset.

ADmad commented:
- Moved `Asset` class under `Routing`
- Updated `UrlHelper` to use `Asset` class
- Added the ability to control the inflection type when inflecting theme/plugin name. I have kept default inflection to `underscore` for now but would like to change it to `dasherize` to consistent with Router.

dereuromark commented on src/Routing/Asset.php:
```diff
@@ -0,0 +1,356 @@
+     * @return void
+     * @psalm-param 'dasherize'|'underscore' $inflectionType
+     */
+    public static function setInflectionType(string $inflectionType)
```
you can set :void here as return type, right?

ADmad replied:
yup, it's an oversight

dereuromark commented on src/Routing/Asset.php:
```diff
@@ -0,0 +1,356 @@
+     * It checks if the plugin is loaded, else filename will stay unchanged for filenames containing dot.
+     *
+     * @param string $name The name you want to plugin split.
+     * @return array Array with 2 indexes. 0 => plugin name, 1 => filename.
```
Isnt this string[] then?

ADmad replied:
No, since plugin can be null

dereuromark commented:
Nice SOLID clean up :)

ADmad committed:
Use "/" as webroot when no request instance available from Router.

ADmad committed:
Update docblocks

dereuromark commented on src/Routing/Asset.php:
```diff
@@ -26,17 +26,17 @@
+     * Set inflection type to use when inflecting plugin/theme name.
      *
-     * @param string $inflectionType Inflection type
+     * @param string $inflectionType Inflection type. Valid values: 'dasherize', 'underscore'.
```
I recommend making those class const then for clarity and to be easier used by dev.

ADmad replied:
Theoretically one can use any `Inflector` method. Someone can use `'humanize'` if they prefer. I don't want to add constants for all `Inflector` methods. `RouterBuilder` methods also have `inflect` option for similar purpose.

I'll change the description to mention the value should be a valid `Inflector` method name.

dereuromark replied:
Cool, ok.

markstory commented:
Looks good :+1: Do you think this is a class we should document in the book? I can write something up in the routing chapter.

ADmad commented:
@markstory Yes please, thanks.

ADmad committed:
Add support for {plugin} placeholder in asset base URLs.

This allows generating correct URLs for plugin/theme assets when using a CDN.

Refs #10172
